### ISSUE TITLE
MSG100 no more working.

### ISSUE THREAD

**Issue by JohannesPl** [bwp91/homebridge-meross#285]
**Describe Your Problem:**

I have: 
MSG100 FW 3.2.2 / HW 3.5.0
iPhone SE2 with IOS14.6
HTTP Catcher 2.8.0

Maybe because I have done a firmwareupdate, but I am no more able to control the MSG100 with the plugin. 

I got the errormessage below, but the IP address is for sure still the same 192.168.178.XX.

When I try to reinstall the plugin I get no "POST" logs as mentioned in the Credential Wiki using HTTP catcher. There are only "CONNECT" entries. Where is my mistake?

**Logs:**

[10.6.2021, 12:52:09] [Meross] MSG100 - Failed to refresh status of Garage: "connect ETIMEDOUT 192.168.178.XX:80" undefined

```
Show the Homebridge logs here. If using Homebridge Config-UI-X, you can easly downlaod your log by visiting the log page on the UI and Clicking the Download Icon in the top right.
Remove any sensitive information, such as yourhomebridge-meross timestamp / sign / messageId / deviceUrl.

Please Turn on Debug Mode before submitting your Issue so more logs can be shared to fix the issue being submitted.

***Read and Paste over this Text the Text Above^^^^^^
```

**Homebridge Config:**
        {
            "name": "Meross",
            "devices": [
                {
                    "name": "Garage",
                    "model": "MSG100",
                    "serialNumber": "ID1234567890",
                    "firmwareRevision": "ID0815",
                    "deviceUrl": "192.168.178.XX",
                    "messageId": "XXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
                    "timestamp": XXXXXXXXXXXX,
                    "sign": "XXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
                    "garageDoorOpeningTime": 20
                }
            ],
            "platform": "Meross"
        },

```json
Paste your homebridge config.json here.
Remove any sensitive information, such as your homebridge-meross timestamp / sign / messageId / deviceUrl.

***Read and Paste over this Text the Text Above^^^^^^
```

**Screenshots:**

**Environment:**

* **Node.js Version**:  14.17.0
* **NPM Version**:  6.14.13
* **Homebridge Version**:  1.3.4
* **Homebridge Meross Plugin Version**: 5.0.2
* **Homebridge Config UI X Plugin Version**: 4.41.1
* **Operating System**: Raspbian 10 buster

**Comment by JohannesPl** [comment]
Ok. Now i got the values with the help of HTTP catcher. 
But the switch is doing nothing when activating via homebridge? It shows, "Garage opening..." and that's all.
The logfiles show nothing, what points to my garage.

**Comment by mag2007** [comment]
Which firmware version are you using? And what is the hardware version?

**Comment by JohannesPl** [comment]
Hi. 

Firmeware-Version is: 3.2.2
Hardware-Version is: 3.5.0

**Comment by starxinca** [comment]
Hi there,
I'm using the same Meross MSG100 garage opener. I just wonder where I could get the **MessageID** please.
I put my iPhone on the same WI-FI and tried HTTP Catcher, however I only got timeStamp and sign. There is no such messageID anywhere. I was so struggling to find it. I appreciate that if you can help me please, even with a screenshot that will very helpful.
One more thing for the **timestamp**, which one should I pick, from Params or Response in HTTP Catcher please.

Firmeware-Version is: 2.1.12
Hardware-Version is: 2.0.0

Many Thanks.

Star

**Comment by CocoaBob** [comment]
I was the author who added the support of MSG100, but now I can't make it work either...

My firmware version is 3.2.3, hardware version is 3.5.0.

I'm afraid Meross has eventually figured out a way to reject the same payload from the same IP address to avoid hacking, that's why we get `socket hang up` error in Homebridge console, which means the connection is terminated by server (in our case, it's terminated by MSG100).

After a bunch of testing, I notice that the same pair of `messageId` & `timestamp` & `sign` can't be used again on the same IP address, but if I copy the payload from `HTTP Catcher` to `Postman` (from iPhone to Mac), then I can use it again, but only once.

I don't think we could figure out the algorithm to generate `sign`, it looks like they are calculating MD5 based on `messageId` and `timestamp`, but we don't know the `salt`...

**Comment by CocoaBob** [comment]
Possible solutions in my mind:
1. I think the firmware has a local database to store the recent `messageId & timestamp & sign` per `IP address`, due to the limited space on the device, there must be a limit number of records. Maybe the list is limited to 100 records, or 1000 records. So if we prepare 1000 pairs of `messageId & timestamp & sign`, and use them one by one, it might work. At least while I was testing, incorrect timestamp still works, the device only checks the `sign`.
2. I tried some websites to reverse the `sign`, but I couldn't succeed. Normally they only have leaked passwords and regular words in their database. If we could reserve one `sign`, we will see the rule of joining `messageId` and `timestamp` and `salt`.
https://md5hashing.net
https://crackstation.net

**Comment by Frankdoc** [comment]
Got the same issue when updating to 3.2.3. Already regret updating :(
And just after the amazon prime deals, could've bought  the homekit version :(

Hope it can be fixed

**Comment by ckuai** [comment]
> Got the same issue when updating to 3.2.3. Already regret updating :(
> And just after the amazon prime deals, could've bought the homekit version :(

I know, but the Refoss version of the same MSG100 HK is having 20% off coupon on Amazon. $40 OTD pricing not too bad.

**Comment by Frankdoc** [comment]
> > Got the same issue when updating to 3.2.3. Already regret updating :(
> > And just after the amazon prime deals, could've bought the homekit version :(
> 
> I know, but the Refoss version of the same MSG100 HK is having 20% off coupon on Amazon. $40 OTD pricing not too bad.

I contacted Meross and they said
1) they can downgrade the firmware from their backend for my account, so I asked them to downgrade it by 1 version
2) they were willing to offer a 20% discount for the Homekit version.

will await the downgrade and report back.

**Comment by ckuai** [comment]
> I contacted Meross and they said
> 
> 1. they can downgrade the firmware from their backend for my account, so I asked them to downgrade it by 1 version
> 2. they were willing to offer a 20% discount for the Homekit version.
> 
> will await the downgrade and report back.

Nice, let us know how the downgrade works. such as does it has the available update shows and you just avoid to update again? or your account is blacklisted for any future upgrade? Thanks. 

BTW, seems the Refoss HK version sold out on amazon already. not sure if it is related to this issue.

**Comment by mrpoune** [comment]
Same here. Just did the firmware upgrade. Dammit!

Please let us know soon how it goes after the downgrade. Was Meross support quick to respond? Did you use the chat option?

Thanks in advance.

**Comment by Frankdoc** [comment]
@CocoaBob 
@mrpoune 
@JohannesPl 

Meross  support replied after +- 10 hours and downgraded me remotely to 3.2.2
After a few minutes it immediately started working again in Homebridge ( without doing any changes).

So easiest option might be just to contact Meross and ask them to downgrade you to 3.2.2.
I used the support form on their website using my meross account username

**Comment by ckuai** [comment]
> Meross support replied after +- 10 hours and downgraded me remotely to 3.2.2
> After a few minutes it immediately started working again in Homebridge ( without doing any changes).

Do you still have the 3.2.3 upgrade available show in the app? or your MSG100 will not get upgrade ever?

> So easiest option might be just to contact Meross and ask them to downgrade you to 3.2.2.
> I used the support form on their website using my meross account username

Thanks, emailed support@meross.com for this.

**Comment by Frankdoc** [comment]
> > Meross support replied after +- 10 hours and downgraded me remotely to 3.2.2
> > After a few minutes it immediately started working again in Homebridge ( without doing any changes).
> 
> Do you still have the 3.2.3 upgrade available show in the app? or your MSG100 will not get upgrade ever?
> 
> > So easiest option might be just to contact Meross and ask them to downgrade you to 3.2.2.
> > I used the support form on their website using my meross account username
> 
> Thanks, emailed [support@meross.com](mailto:support@meross.com) for this.

No it's showing 'update available', so if I want I could update again.

**Comment by ckuai** [comment]
> Meross support replied after +- 10 hours and downgraded me remotely to 3.2.2
> After a few minutes it immediately started working again in Homebridge ( without doing any changes).

Hi, I also contacted Meross support and got downgraded to 3.2.1 but I am still having issue. 
```
[7/7/2021, 12:30:18 AM] [Meross] MSG100 - Failed to refresh status of Garage Door: "socket hang up" undefined
```
which homebridge-meross plugin version you are on? I was on 2021 Feb 3.5.0 version but updated to latest 5.0.3 due to this issue. I also tried to downgrade to 3.5.0 plugin, but the issue remain the same. Not sure if I should ask to down grade to 3.2.2 instead. but 3.2.1 should work too.

**Comment by Frankdoc** [comment]
> > Meross support replied after +- 10 hours and downgraded me remotely to 3.2.2
> > After a few minutes it immediately started working again in Homebridge ( without doing any changes).
> 
> Hi, I also contacted Meross support and got downgraded to 3.2.1 but I am still having issue.
> 
> ```
> [7/7/2021, 12:30:18 AM] [Meross] MSG100 - Failed to refresh status of Garage Door: "socket hang up" undefined
> ```
> 
> which homebridge-meross plugin version you are on? I was on 2021 Feb 3.5.0 version but updated to latest 5.0.3 due to this issue. I also tried to downgrade to 3.5.0 plugin, but the issue remain the same. Not sure if I should ask to down grade to 3.2.2 instead. but 3.2.1 should work too.

FW 3.2.2 Meross
HB plugin 3.5.0 ( also had issues with the v5 release, so I went back). Your config file is Ok since version 5 is different?

**Comment by ituri** [comment]
FYI, I experienced all the problems above and asked support to downgrade me to 3.2.2 which they did after a while. Interestingly enough, they also wrote this in their reply:
>Btw, if you were talking about the meross plugin on github, you may contact the author to modify the coding part then you all can use the new firmware again.

I asked them about the salt. If they actually provide this information (waiting for their reply right now), maybe there's a chance for us to get it working again with the latest firmware.

**Comment by CocoaBob** [comment]
@ituri Wow, thanks for your information! The `salt` thing was my guess, actually we need to know the entire algorithm/way to generate the `sign`. The process might look like this

[INPUT]
- `messageId`: a random 128bit uuid
- `timestamp`: current timestamp
- `salt`: maybe it's there, maybe not

[CALCULATION]
- Join the above inputs in a certain way, e.g. `messageId`+`timestamp`+`salt`, or `salt`+`messageId`+`timestamp`
- Use MD5 or similar algorithm to hash it

[OUTPUT]
- `sign`: a 128-bit hash generated by MD5 or similar algorithm

I doubt if they could provide us this kind of info, it's dangerous for them. But on the other hand, we may find a jailbroken iPhone and reverse engineer their iOS app 🤐(or Android app maybe that's easier)

**Comment by ituri** [comment]
So I received an answer to my question:
> As far as I know the calculation is not changed, the new firmware fixed the problem that messageId can be replayed and attacked.

There was actually an update to the iOS app tonight, which might also include the new firmware support is talking about. Anyone brave enough to try the new update? 😊

**Comment by CocoaBob** [comment]
> fixed the problem that messageId can be replayed and attacked

That's why our plugin stops working, if they are going to fix it on all products, then our entire Meross plugin will stop working 🤦‍♂️

**Comment by NexGen-3D-Printing** [comment]
Just an FYI, I have the Homekit version of the MSG100, it has the latest software version 4.1.14 and hardware version 4.0.0 which is the latest and works great with this plugin.

And for those that are thinking that you don't need this plugin with the Homekit version, you are wrong, as Apple or Meross have taken away any real usability of the device, it functions much better through Homebribge.

Native Homekit version:

- Cant use any Automation to open the Garage Door
- Cant assign any smart buttons or switches to open the Garage Door
- Cant use any kind of dummy device to open the Garage Door
- Cant use the Homebridge Alexa plugin to open the Garage Door
- Cant use the device natively in Alexa outside of the USA
- You can only open the Garage Door from inside the Meross or Home app with the native buttons provided or via Siri from an unlocked and authorised IOS device.

But once added to this plugin you can do what ever you like with it, I had been on a quest to purchase more authentic Homekit devices, but after this experience I think I will be looking for devices that only function under Homebridge.

Long story short, the Homekit version works great with HTTP Catcher, I will make sure not to upgrade the firmware now as it functions perfectly and I prefer it stay that way.

**Comment by bizzle12368239** [comment]
Same issue. Definitely the firmware update causing it. I have asked Meross to downgrade my firmware, hopefully they do it like they've done for a few others. 

Edit: they answered within the day and downgraded me. Now it works again. Happy to upgrade the firmware in the future if we think it is possible to get it working with home bridge.

**Comment by admoseremic** [comment]
I asked Meross to downgrade my firmware to 3.2.2 and it worked for me.

**Comment by charlieschmidt** [comment]
Another voice/datapoint on this, for the time being - I asked them to downgrade my two devices - they took maybe 48hrs to do it and everything was back to normal.

As I told them - holy shit that's some great service.  "Here, let us support you do this unsupported thing that we sell another device for anyway."

**Comment by mag2007** [comment]
So I contacted them too, after I upgraded my opener. They downgraded it the next days, so I am back on version 3.2.2 and HB Plugin version 3.5.0. But still facing issue with:
`Failed to POST to the Meross Device MSG100 at 192.160.68.109: Error: connect ETIMEDOUT 192.160.68.109:80
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1148:16) {
  errno: -110,
  code: 'ETIMEDOUT',
  syscall: 'connect',
  address: 'ip of garage door opener',
  port: 80
}`

But I could not get it to work until update too, so I think there is something else. Do you guys know, if the http catcher still works with latests iOS App? They "fixed" something, as much I read.

**Comment by bizzle12368239** [comment]
Yes HTTP catcher works fine even on ios 15 beta, so best to start from
scratch and set it up again.

**Comment by mag2007** [comment]
Can you provide full config of your setup? What is missing here? HB Plugin version 3.5.0.

```
{
            "name": "Garage",
            "model": "MSG100",
            "serialNumber": "RANDOM NUMBERS",
            "firmwareRevision": "3.2.2",
            "deviceUrl": "DEVICE IP",
            "channel": 0,
            "messageId": "GOT FROM HTTP CATCHER",
            "timestamp": GOT FROM HTTP CATCHER,
            "sign": "GOT FROM HTTP CATCHER",
            "garageDoorOpeningTime": 15,
            "accessory": "Meross"
        },
```

**Comment by Frankdoc** [comment]
When downgraded: did you also notice that google home isn't working anymore? When I tap the garage door it doesn't let me do anything. Previously I could open/close or ask google assistant to open/close garage door

**Comment by CocoaBob** [comment]
💡🤫🤐

**Comment by ituri** [comment]
Is it safe to upgrade the device's firmware, now that the plugin has switched to Meross' cloud?

**Comment by bwp91** [comment]
> Is it safe to upgrade the device's firmware, now that the plugin has switched to Meross' cloud?

The plugin hasn't 'switched' to meross cloud, the option has been added, but you can continue to use local devices (just with different config params. I _think_ this new method has made garage doors work again on the latest firmware, but I can't be certain as I don't own the device.

Hopefully someone else can answer that for you!

**Comment by CocoaBob** [comment]
I'm using the latest firmware on my MSG100, and I just updated to v6.5.0, it's working (of cause 😉).

**Comment by bwp91** [comment]
anyone here with an MSG100 device? would you have any insight into this issue?
https://github.com/bwp91/homebridge-meross/issues/512
